### ISSUE TITLE
NullPointerException - copyObject appears to be broken

### ISSUE THREAD

**Issue by tawrahim** [minio/minio-java#556]
It appears that `copyObject` method provided does not work. Here is the exception that is being thrown

````
java.lang.NullPointerException
	at io.minio.Signer.setSigningKey(Signer.java:199)
	at io.minio.Signer.signV4(Signer.java:231)
	at io.minio.MinioClient.execute(MinioClient.java:811)
	at io.minio.MinioClient.executePut(MinioClient.java:1083)
	at io.minio.MinioClient.copyObject(MinioClient.java:1578)
````

I did some research and noticed that others are also having this issue. You can take a look at [this repo](https://github.com/jeppe-style/performance-test-manager/blob/5500e6c2f3a4bb39a593c3c282eb9662582669b1/application/src/main/java/cloud/benchflow/testmanager/services/external/MinioService.java#L254) to confirm that others have this issue.

**Comment by harshavardhana** [comment]
> java.lang.NullPointerException
> 	at io.minio.Signer.setSigningKey(Signer.java:199)
> 	at io.minio.Signer.signV4(Signer.java:231)
> 	at io.minio.MinioClient.execute(MinioClient.java:811)
> 	at io.minio.MinioClient.executePut(MinioClient.java:1083)
> 	at io.minio.MinioClient.copyObject(MinioClient.java:1578)

Do you have a full sample code on this? we have functional tests and we are not able to reproduce this on our end.. 

Can you try to reproduce this with latest master? @tawrahim

**Comment by tawrahim** [comment]
@harshavardhana - I created a repo to reproduce the bug - [Link to to repo](https://github.com/tawrahim/minio-copy-object-bug) 

When you say with latest mater do you mean **minio-3.0.3-all.jar** ?

**Comment by balamurugana** [comment]
@tawrahim The example code is quite straight forward which is similar to the functional test. I don't see any problem with that.

There could be below problems your side.
1. minio java requires below dependencies.  Make sure you have correct version of them.
```
        compile "com.google.http-client:google-http-client-xml:1.20.0"
        compile "com.google.guava:guava:18.0"
        compile "com.squareup.okhttp:okhttp:2.7.2"
        compile "com.squareup.okio:okio:1.6.0"
        compile "joda-time:joda-time:2.7"
        compile "com.fasterxml.jackson.core:jackson-annotations:2.8.4"
        compile "com.fasterxml.jackson.core:jackson-core:2.8.4"
        compile "com.fasterxml.jackson.core:jackson-databind:2.8.4"
        compile 'com.google.code.findbugs:annotations:3.0.1'
        compile 'com.google.code.findbugs:jsr305:3.0.1'
```
2. minio-java requires Java 1.7 or 1.8.  Please check what version of java you use.

I see you are using minio server.  Please provide me 
1. output of `minio version` command.
2. Run your example code and provide me full stacktrace.
3. Point your example code to `play.minio.io` and check whether its working or not.

**Comment by tawrahim** [comment]
You are right - It is actually the same as the functional test

```sh
/go/src/app # minio version
Version: 2017-01-25T03:14:52Z
Release-Tag: RELEASE.2017-01-25T03-14-52Z
Commit-ID: f8e4700a11065967242b8857045ce7dee607722d
/go/src/app # 
```
Here is the full stack trace

```java
Exception in thread "main" java.lang.NullPointerException
	at io.minio.Signer.setSigningKey(Signer.java:199)
	at io.minio.Signer.signV4(Signer.java:231)
	at io.minio.MinioClient.execute(MinioClient.java:811)
	at io.minio.MinioClient.executePut(MinioClient.java:1083)
	at io.minio.MinioClient.copyObject(MinioClient.java:1578)
	at io.minio.MinioClient.copyObject(MinioClient.java:1470)
	at CopyObject.main(CopyObject.java:72)
```

**Comment by balamurugana** [comment]
@tawrahim Thanks for the info.

Can you confirm dependencies and `play.minio.io`?

**Comment by tawrahim** [comment]
For `play.minio.io` which **access id** and **secret key** should I be using?

**Comment by tawrahim** [comment]
Also for the dependencies - I am just using the `minio-3.0.3-all.jar` which I believe contains all the deps.

**Comment by balamurugana** [comment]
@tawrahim If you are using `minio-3.0.3-all.jar` then its good to go. Below is `play.minio.io`  detail.

```java
      /* play.minio.io for test and development. */
      MinioClient minioClient = new MinioClient("https://play.minio.io:9000", "Q3AM3UQ867SPQQA43P2F",
                                                "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG");
```

**Comment by tawrahim** [comment]
Looks like `play.minio.io` is down

```java
Exception in thread "main" java.net.SocketTimeoutException: connect timed out
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at com.squareup.okhttp.internal.Platform.connectSocket(Platform.java:101)
	at com.squareup.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:137)
	at com.squareup.okhttp.internal.io.RealConnection.connect(RealConnection.java:108)
	at com.squareup.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:184)
	at com.squareup.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
	at com.squareup.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
	at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281)
	at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
	at com.squareup.okhttp.Call.getResponse(Call.java:286)
	at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:243)
	at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:205)
	at com.squareup.okhttp.Call.execute(Call.java:80)
	at io.minio.MinioClient.execute(MinioClient.java:826)
	at io.minio.MinioClient.executePut(MinioClient.java:1083)
	at io.minio.MinioClient.makeBucket(MinioClient.java:2135)
	at io.minio.MinioClient.makeBucket(MinioClient.java:2100)
	at CopyObject.main(CopyObject.java:64)
```

I actually cannot even get to **https://play.minio.io:9000/** in the browser

**Comment by harshavardhana** [comment]
> I actually cannot even get to https://play.minio.io:9000/ in the browser

Play seems to be working fine, looks like you have a network problem. 

```
 mc ls play/testbucket
[2017-04-16 22:35:27 PDT] 136KiB filename.txt
[2017-04-15 11:58:13 PDT]  13MiB my-objectname
[2017-04-16 16:24:57 PDT]     6B test1
[2017-04-16 16:07:28 PDT]     4B test1.txt
[2017-04-16 16:07:11 PDT]     4B test2.txt
[2017-04-18 09:07:29 PDT]     0B 1/
[2017-04-18 09:07:29 PDT]     0B accounts/
[2017-04-18 09:07:29 PDT]     0B documents/
[2017-04-18 09:07:29 PDT]     0B resumes/
```

**Comment by tawrahim** [comment]
It actually looks like it was because I am behind a coprate firewall. I just switched to my hotspot and it looks like I am able to reach `play.minio.io`

I still saw the exception testing against play.minio.io

```java
my-objectname is uploaded successfully
Exception in thread "main" java.lang.NullPointerException
	at io.minio.Signer.setSigningKey(Signer.java:199)
	at io.minio.Signer.signV4(Signer.java:231)
	at io.minio.MinioClient.execute(MinioClient.java:811)
	at io.minio.MinioClient.executePut(MinioClient.java:1083)
	at io.minio.MinioClient.copyObject(MinioClient.java:1578)
	at io.minio.MinioClient.copyObject(MinioClient.java:1470)
	at CopyObject.main(CopyObject.java:75)
```

**Comment by tawrahim** [comment]
I am going to update my repo with the play URL so that you guys can test it directly.

**Comment by balamurugana** [comment]
@tawrahim I am able to reproduce this issue with `minio-java-3.0.3-all.jar`.  However it works fine with latest development master.

Can you confirm whether minio-java master works for you?

**Comment by balamurugana** [comment]
@tawrahim PR https://github.com/minio/minio-java/pull/519 fixes the problem.

**Comment by tawrahim** [comment]
Glad that you were able to reproduce this. Do you think you can drop me a jar of master that I can try with? Also when do you think it is going to make it into maven? Our current project pulls directly from maven central.

**Comment by harshavardhana** [comment]
> Glad that you were able to reproduce this. Do you think you can drop me a jar of master that I can try with? Also when do you think it is going to make it into maven? Our current project pulls directly from maven central.

We will make a release @tawrahim

**Comment by harshavardhana** [comment]
Fixed by #519 - also released 3.0.4
